src/main/java/htsjdk/samtools/CSIIndex.java

```diff
@@ -0,0 +1,466 @@
+package htsjdk.samtools;
+
+import htsjdk.samtools.seekablestream.SeekableStream;
+import htsjdk.samtools.util.RuntimeIOException;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.*;
+
+
+public class CSIIndex extends AbstractBAMFileIndex implements BrowseableBAMIndex {
+
+    private int binDepth;
+    private int minShift;
+    private int maxBins;
+    private int maxSpan;
+    private byte[] auxData;
+    private int nReferences;
+    private int metaDataPos = -1;
+
+    /**
+     * Constructors
+     */
+
+    public CSIIndex(final SeekableStream stream, final SAMSequenceDictionary dictionary) {
+        this(new IndexStreamBuffer(stream), stream.getSource(), dictionary);
+    }
+
+    public CSIIndex(final Path path, boolean enableMemoryMapping, final SAMSequenceDictionary dictionary) {
+        this(IndexFileBufferFactory.getBuffer(path.toFile(), enableMemoryMapping), path.toString(), dictionary);
```

Comment:
It's not safe to do path.toFile because many paths will not be representable as files.  This constructor should create a SeekableStream from the path instead.  And it probably shouldn't have the memory mapping parameter after the change since I don't believe SeekableStream respects it.